apps/emqx_resource/test/emqx_resource_SUITE.erl

```diff
@@ -2291,6 +2291,67 @@ t_expiration_retry_batch_multiple_times(_Config) ->
     ),
     ok.
 
+t_recursive_flush(_Config) ->
+    emqx_connector_demo:set_callback_mode(async_if_possible),
+    {ok, _} = emqx_resource:create(
+        ?ID,
+        ?DEFAULT_RESOURCE_GROUP,
+        ?TEST_RESOURCE,
+        #{name => test_resource},
+        #{
+            query_mode => async,
+            batch_size => 1,
+            batch_time => 10_000,
+            worker_pool_size => 1
+        }
+    ),
+    do_t_recursive_flush().
+
+t_recursive_flush_batch(_Config) ->
+    emqx_connector_demo:set_callback_mode(async_if_possible),
+    {ok, _} = emqx_resource:create(
+        ?ID,
+        ?DEFAULT_RESOURCE_GROUP,
+        ?TEST_RESOURCE,
+        #{name => test_resource},
+        #{
+            query_mode => async,
+            batch_size => 2,
+            batch_time => 10_000,
+            worker_pool_size => 1
+        }
+    ),
+    do_t_recursive_flush().
+
+do_t_recursive_flush() ->
+    ?check_trace(
+        begin
+            Timeout = 1_000,
+            Pid = spawn_link(fun S() ->
+                emqx_resource:query(?ID, {inc_counter, 1}),
+                S()
+            end),
+            %% we want two reflushes to happen before we analyze the
+            %% trace, so that we get a single full interaction
+            {ok, _} = snabbkaffe:block_until(
+                ?match_n_events(2, #{?snk_kind := buffer_worker_flush_ack_reflush}), Timeout
+            ),
+            unlink(Pid),
+            exit(Pid, kill),
+            ok
+        end,
+        fun(Trace) ->
+            %% check that a recursive flush leads to a new call to flush/1
+            Pairs = ?find_pairs(
+                #{?snk_kind := buffer_worker_flush_ack_reflush},
+                #{?snk_kind := buffer_worker_flush},
+                Trace
+            ),
+            ?assert(lists:any(fun(E) -> E end, [true || {pair, _, _} <- Pairs]))
```

Comment:
Just a note: from what I understand, it seems that there are a few sources for flushing that could lead to a `buffer_worker_flush` event, such as the return from the async call and the flush timer, so these could also give the impression that the "reflush" causes the `flush_worker` call.  

But I can't see a stronger way to assert that causality at this moment. :thinking: 

Did the `{internal, flush}` event cause problems in the past?  That could be a way to differentiate the cause (by the way, `running/3` with `internal, flush, _` is now unreachable).

I remember there was a [situation](https://github.com/emqx/emqx/blob/056bc71af26d486b220cb5780151e1d8ecef3f76/apps/emqx_resource/src/emqx_resource_buffer_worker.erl#L199-L203) when returning the internal event could crash, but it seemed to be resolved as the flush was called later via a timer.  :thinking:

Reply:
Thank you for your pointers! It was hard finding a good way in general to test this.

Stone and I found a crash while testing the influx bridge when `[{next_event, internal, flush}]}` was triggered. We simply patched it up and I have not thought about any deeper implications.

If it is okay, I'll go ahead and merge it.